flow/table.go

```diff
@@ -0,0 +1,157 @@
+package flow
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/gopacket"
+
+	"github.com/redhat-cip/skydive/logging"
+)
+
+type FlowTable struct {
+	lock  sync.RWMutex
+	table map[string]*Flow
+}
+
+type FlowTableAsyncNotificaionUpdate interface {
+	AsyncNotificaionUpdate(every time.Duration)
+}
+
+func NewFlowTable() *FlowTable {
+	return &FlowTable{table: make(map[string]*Flow)}
+}
+
+func (ft *FlowTable) String() string {
+	return fmt.Sprintf("%d flows", len(ft.table))
+}
+
+func (ft *FlowTable) Update(flows []*Flow) {
+	ft.lock.Lock()
+	defer ft.lock.Unlock()
+	for _, f := range flows {
+		_, found := ft.table[f.UUID]
+		if !found {
+			ft.table[f.UUID] = f
+		} else if f.UUID != ft.table[f.UUID].UUID {
+			logging.GetLogger().Error("FlowTable Collision ", f.UUID, ft.table[f.UUID].UUID)
+		}
+	}
+}
+
+type ExpireFunc func(f *Flow)
+
+func (ft *FlowTable) AsyncExpire(fn ExpireFunc, every time.Duration) {
+	ticker := time.NewTicker(every)
+	defer func() {
+		ticker.Stop()
+	}()
+	for {
+		select {
+		case now := <-ticker.C:
+			flowTableSzBefore := len(ft.table)
+			expire := now.Unix() - int64((every).Seconds())
+
+			ft.lock.Lock()
+			defer ft.lock.Unlock()
```

Comment:
Dead lock here, defer will be applied when leaving the function, but since we are in a infinite loop...